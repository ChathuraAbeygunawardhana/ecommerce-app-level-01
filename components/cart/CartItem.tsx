import React from 'react';
import { View, Text, Image, TouchableOpacity, StyleSheet } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import sampleData from '../../assets/sample.json';

interface CartItemProps {
  item: {
    id: string;
    name: string;
    price: number;
    image: string;
    quantity: number;
  };
  currentColors: {
    background_01: string;
    background_02: string;
    text: string;
    grey: string;
  };
  increaseQuantity: (id: string) => void;
  decreaseQuantity: (id: string) => void;
  removeFromCart: (id: string) => void;
}

const CartItem: React.FC<CartItemProps> = ({
  item,
  currentColors,
  increaseQuantity,
  decreaseQuantity,
  removeFromCart,
}) => {
  const product = sampleData.find(p => p.id === item.id);

  return (
    <View style={[styles.item, { backgroundColor: currentColors.background_02 }]}>
      <Image source={{ uri: item.image }} style={styles.image} />
      <View style={styles.details}>
        <Text style={[styles.name, { color: currentColors.text }]} numberOfLines={1}>
          {item.name.split(' ').slice(0, 3).join(' ').substring(0, 15)}
              {/* TODO: These kind of small logics should move to utility functions */}
        </Text>
        <Text style={[styles.price, { color: currentColors.grey }]}>
          ${item.price}
        </Text>
        <View style={styles.quantityContainer}>
          <TouchableOpacity onPress={() => decreaseQuantity(item.id)} disabled={item.quantity === 1}>
            <View style={[styles.iconContainer, { backgroundColor: currentColors.background_01 }]}>
              <Ionicons name="remove" size={15} color={currentColors.text} />
            </View>
          </TouchableOpacity>
          <Text style={[styles.quantity, { color: currentColors.text }]}>
            {item.quantity > 0 ? item.quantity : 'Out of Stock'}
          </Text>
          <TouchableOpacity onPress={() => {
            if (product && item.quantity < product.quantity) {
              increaseQuantity(item.id);
            }
          }} disabled={product && item.quantity >= product.quantity}>
            <View style={[styles.iconContainer, { backgroundColor: Colors.lightBlue }]}>
              <Ionicons name="add" size={15} color={Colors.white} />
            </View>
          </TouchableOpacity>
        </View>
        <TouchableOpacity onPress={() => removeFromCart(item.id)} style={styles.deleteButton}>
          <Ionicons name="trash-outline" size={24} color={currentColors.grey} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 16,
    marginVertical: 8,
    marginHorizontal: 16,
    borderRadius: 8,
  },
  image: {
    width: 80,
    height: 80,
    borderRadius: 8,
    marginRight: 16,
  },
  details: {
    flex: 1,
  },
  name: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 8,
  },
  price: {
    fontSize: 16,
    marginBottom: 8,
  },
  quantityContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 8,
  },
  quantity: {
    fontSize: 16,
    marginHorizontal: 10,
  },
  iconContainer: {
    width: 24,
    height: 24,
    borderRadius: 12,
    justifyContent: 'center',
    alignItems: 'center',
  },
  deleteButton: {
    position: 'absolute',
    top: 3,
    right: 2,
  },
});

export default CartItem;
