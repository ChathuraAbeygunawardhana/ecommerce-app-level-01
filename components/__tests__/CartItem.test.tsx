import React from 'react';
import renderer from 'react-test-renderer';
import CartItem from '../Cart/CartItem';

describe('CartItem', () => {
  it('renders correctly', () => {
    const item = {
      id: '1',
      name: 'Product Name',
      price: 100,
      image:
        'https://s3-eu-west-1.amazonaws.com/api.themeshplatform.com/media/7e386191b2ee40b290886a05d3e10e24_nike-air-relentless-a.jpg',
      quantity: 1,
    };
    const tree = renderer
      .create(
        <CartItem
          item={item}
          removeFromCart={() => {}}
          increaseQuantity={() => {}}
          decreaseQuantity={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
