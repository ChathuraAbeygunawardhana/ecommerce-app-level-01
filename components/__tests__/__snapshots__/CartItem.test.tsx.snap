// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "borderBottomColor": "#ddd",
      "borderBottomWidth": 1,
      "borderRadius": 8,
      "elevation": 2,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginHorizontal": 16,
      "marginVertical": 8,
      "padding": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 8,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://example.com/image.jpg",
      }
    }
    style={
      {
        "borderRadius": 8,
        "height": 80,
        "marginRight": 16,
        "width": 80,
      }
    }
  />
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 16,
          "fontWeight": "bold",
          "marginBottom": 8,
        }
      }
    >
      Product Name
    </Text>
    <Text
      style={
        {
          "color": "#666",
          "fontSize": 16,
          "marginBottom": 8,
        }
      }
    >
      $
      100
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 8,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#ddd",
              "borderRadius": 12,
              "height": 24,
              "justifyContent": "center",
              "width": 24,
            }
          }
        >
          <Text />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#666",
            "fontSize": 16,
            "marginHorizontal": 10,
          }
        }
      >
        1
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#ddd",
              "borderRadius": 12,
              "height": 24,
              "justifyContent": "center",
              "width": 24,
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "position": "absolute",
          "right": 8,
          "top": 8,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
